src/main/scala/rise/core/DSL/infer.scala

```diff
@@ -0,0 +1,244 @@
+package rise.core.DSL
+
+import Type.freshTypeIdentifier
+import rise.core._
+import rise.core.traversal.{Continue, Result, Stop}
+import rise.core.types.InferenceException.error
+import rise.core.types._
+
+import scala.collection.mutable
+
+object infer {
+  private [DSL] def apply(e: Expr,
+            explDep: Flags.ExplicitDependence = Flags.ExplicitDependence.Off): Expr = {
+    val constraints = mutable.ArrayBuffer[Constraint]()
+    val (typed_e, ftvSubs) = constrainTypes(e, constraints, mutable.Map())
+    val solution = unfreeze(ftvSubs, Constraint.solve(constraints.toSeq, Seq())(explDep))
+    traversal.DepthFirstLocalResult(typed_e, Visitor(solution))
```

Comment:
will this DSL inference perform something like `printTypeAndTypeHoles`? https://github.com/rise-lang/shine/pull/76/files#diff-c9cc95d0ecff6435a10227129d923a27d7cc9eb79f2026c9d7f2835fc0b9c0d0L31

Reply:
does this DSL enforce unique names for DPIA? https://github.com/rise-lang/shine/pull/76/files#diff-c9cc95d0ecff6435a10227129d923a27d7cc9eb79f2026c9d7f2835fc0b9c0d0L21

Reply:
I do not know the relationship to `printTypeAndTypeHoles` (neither do I understand what is done there).

We are not enforcing unique names in the type inference process.
Substitution is no longer relying on unique names.

Reply:
`printTypeAndTypeHoles` will handle debug primitives:
- print the types of the `printType` primitives
- print the types of the `typeHole` primitives, throwing an error if a type hole is found (because a type hole makes the expression incomplete)

I think that we should add this to the new DSL type inference (while avoiding excessive prints), or alternatively move this code to `fromRise` on the transition to DPIA.

Reply:
About unique names, I think that we still require unique names on the DPIA side, so we might want to move the enforce to `fromRise` if the type inference does not require it anymore.